File: zig/tests/binary_search_test.zig
const std = @import("std");
const testing = std.testing;
const binarySearch = @import("../src/binary_search.zig").binarySearch;

test "binary search" {
    // Test with target in middle
    try testing.expectEqual(@as(?usize, 2), binarySearch(i32, &[_]i32{ 1, 2, 3, 4, 5 }, 3));

    // Test with target at start
    try testing.expectEqual(@as(?usize, 0), binarySearch(i32, &[_]i32{ 1, 2, 3, 4, 5 }, 1));

    // Test with target at end
    try testing.expectEqual(@as(?usize, 4), binarySearch(i32, &[_]i32{ 1, 2, 3, 4, 5 }, 5));

    // Test with target not present
    try testing.expectEqual(@as(?usize, null), binarySearch(i32, &[_]i32{ 1, 2, 3, 4, 5 }, 6));

    // Test with empty array
    try testing.expectEqual(@as(?usize, null), binarySearch(i32, &[_]i32{}, 1));

    // Test with single element
    try testing.expectEqual(@as(?usize, 0), binarySearch(i32, &[_]i32{1}, 1));

    // Test with negative numbers
    try testing.expectEqual(@as(?usize, 1), binarySearch(i32, &[_]i32{ -5, -3, -1, 0, 2 }, -3));

    // Test with duplicate elements (should return first occurrence)
    try testing.expectEqual(@as(?usize, 1), binarySearch(i32, &[_]i32{ 1, 2, 2, 2, 3 }, 2));
}

File: ts/tests/binary_search.test.ts
import { binarySearch } from '../src/binary_search';

describe('binarySearch', () => {
    test('should find target in middle of array', () => {
        expect(binarySearch([1, 2, 3, 4, 5], 3)).toBe(2);
    });

    test('should find target at start of array', () => {
        expect(binarySearch([1, 2, 3, 4, 5], 1)).toBe(0);
    });

    test('should find target at end of array', () => {
        expect(binarySearch([1, 2, 3, 4, 5], 5)).toBe(4);
    });

    test('should return -1 when target not present', () => {
        expect(binarySearch([1, 2, 3, 4, 5], 6)).toBe(-1);
    });

    test('should handle empty array', () => {
        expect(binarySearch([], 1)).toBe(-1);
    });

    test('should handle single element array', () => {
        expect(binarySearch([1], 1)).toBe(0);
    });

    test('should handle negative numbers', () => {
        expect(binarySearch([-5, -3, -1, 0, 2], -3)).toBe(1);
    });

    test('should handle duplicate elements (returns first occurrence)', () => {
        expect(binarySearch([1, 2, 2, 2, 3], 2)).toBe(1);
    });

    test('should handle null or undefined array', () => {
        expect(binarySearch(null as unknown as number[], 1)).toBe(-1);
        expect(binarySearch(undefined as unknown as number[], 1)).toBe(-1);
    });
});
